feat(perf): Document span operations spec (#452)
getsentry/develop: src/docs/sdk/performance/span-operations.mdx

---
title: "Span Operations"
---

Span operations are a short code identifying the type of operation the span is measuring. Span operations are low cardinality attributes - they should be as general as possible while still being human readable and useful.

Operations are expected to follow [OpenTelemetry's semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/README.md) as much as possible.

Below is a table containing a list of span operations that are recommended to be used.

> Operations are expected to generally take the form: `CATEGORY.ACTION`, where the categories and actions can have optional identifiers. For example, a redis database operation takes the form `db.sql.query`, where `db` is the category, `sql` is the category identifier, and `query` is the action. Another example is a mobile app rendering it's screen. It would have the operation `ui.load.hot`, where `ui` is the category, `load` is the action, and `hot` is the action identifier.

Author: Yeah we can decide on how to make individual platforms more consistent after we come to a general decision on the spec here.

> Operations are expected to generally take the form: `CATEGORY.ACTION`, where the categories and actions can have optional identifiers. For example, a redis database operation takes the form `db.sql.query`, where `db` is the category, `sql` is the category identifier, and `query` is the action. Another example is a mobile app rendering it's screen. It would have the operation `ui.load.hot`, where `ui` is the category, `load` is the action, and `hot` is the action identifier.

Reviewer: > I think Redis was an unfortunate example, it doesn't use SQL.

The Redis example also doesn't follow the `CATEGORY.ACTION` format we're just starting to introduce in the prose, so I'd suggest starting with something simpler, and following up with a more complex case.

```suggestion
Operations are expected to generally take the form: `CATEGORY.ACTION`, where the categories and actions can have optional identifiers. For example, an instrumented HTTP server creates transaction spans named `http.server`, where `http` is the category, and `server` is the category identifier. Another example is a mobile app rendering a screen with a span operation `ui.load.hot`, where `ui` is the category, `load` is the action, and `hot` is the action identifier.
```

Hmm not happy with `http.server`... I'd rather have something that follows the `CATEGORY.ACTION` format there.

> Operations are expected to generally take the form: `CATEGORY.ACTION`, where the categories and actions can have optional identifiers. For example, a redis database operation takes the form `db.sql.query`, where `db` is the category, `sql` is the category identifier, and `query` is the action. Another example is a mobile app rendering it's screen. It would have the operation `ui.load.hot`, where `ui` is the category, `load` is the action, and `hot` is the action identifier.

Reviewer: 🤔 💭 

There's both "category identifier" and "action identifier" on top of "category" and "action", but in the current shape of the spec we only see 1, 2 or 3 dot-separated items making up the operation name.

I wonder if there's a simpler way to refer to the parts. `CATEGORY[.CAT_IDENT].ACTION[.ACTION_IDENT]` is ambiguous to parse (for computers and humans), and so the 4 names for the parts feel like unnecessary complexity that we can't really make good use of.

For example, we cannot split up the `span.op` and render something descriptive in Sentry like "Operation: DB; Action: Insert".

It's important to keep categories consistent between sdks and integrations as they are used by Sentry in the [Operations Breakdown](https://docs.sentry.io/product/sentry-basics/tracing/event-detail/#operations-breakdown) feature. For example, both `db.init` and `db.query` will be categorized as database operations. The default operations breakdown config can be seen [here](https://github.com/getsentry/sentry/blob/809b7fe54c6f06cc1e4c503cf83ded896472a011/src/sentry/projectoptions/defaults.py#L74).

Having both a category and action is recommended. Operations should be lowercased.

# List of Operations

## Browser

> | Op         | Usage           | Description                                                                                                         |

Author: Yeah this is a function of this being WIP. Will rename `Op` -> `Category` to make my intentions here a little more clear. 

The `Usage` column was just a collection of operations I've seen being used throughout the SDKs

| navigation |                 | Client-side browser history change in a web application                                                             |
| resource   |                 | Resource as per [Performance Resource Timing](https://w3c.github.io/resource-timing/#sec-performanceresourcetiming) |
|            | resource.script |                                                                                                                     |
|            | resource.link   |                                                                                                                     |
|            | resource.css    |                                                                                                                     |
|            | resource.img    |                                                                                                                     |
| browser    |                 | Usage of browser APIs or functionality                                                                              |

> |            | browser.paint   |                                                                                                                     |

Author: For sure. We've been using plain categories (with no actions or identifiers) for now. We can change this, but I think that's something we can come back to.

![image](https://user-images.githubusercontent.com/18689448/140058184-39acf5fa-9fdc-48f3-8e88-bb7b9b1d9cb1.png)

| mark       |                 | Usage of [performance.mark() API](https://developer.mozilla.org/en-US/docs/Web/API/Performance/mark)                |
| measure    |                 | Usage of [performance.measure() API](https://developer.mozilla.org/en-US/docs/Web/API/Performance/measure)          |
| user       |                 |                                                                                                                     |
| animate    |                 |                                                                                                                     |

> ### JS Frameworks

Author: Yes I think the natural next step here is to build out a set of semantic conventions for ourselves on how you would build different integrations and what type of info (error -> context, tags, capture calls, performance -> transactions, spans, tags) you collect.

I think for now we keep this here as a starting point.

> | Op      | Usage        | Description                                    |

Author: Added as part of Browser operations

| react   |              | Spans related to [React](https://reactjs.org/) |
|         | react.mount  |                                                |
|         | react.render |                                                |
| vue     |              |                                                |
|         | vue.mount    |                                                |
|         | vue.update   |                                                |
| angular |              |                                                |
| ember   |              |                                                |

## Web Server

| Op        | Usage           | Description |
| --------- | --------------- | ----------- |
| http      |                 |             |
|           | http.client     |             |
|           | http.server     |             |
| rpc       |                 |             |
| grpc      |                 |             |
| template  |                 |             |
|           | template.init   |             |
|           | template.parse  |             |
|           | template.render |             |

> | view      |                 |             |

Author: I think I might use `render` instead then.

|           | view.render     |             |
| serialize |                 |             |

## Web Frameworks

| Op      | Usage             | Description |
| ------- | ----------------- | ----------- |
| django  |                   |             |
|         | django.middleware |             |
|         | django.view       |             |
| express |                   |             |
| rails   |                   |             |
| rack    |                   |             |

> | console |                   |             |

Author: grabbed from https://github.com/getsentry/sentry-symfony/blob/5129f962a9b921f08fa2100fc5593cd6d8f06566/src/EventListener/TracingConsoleListener.php#L62, but this seems to be more related to running stuff on the command line. I'll move into web server (since using rails, django cmd line is pretty common operations).

## Database

| Op  | Usage            | Description                |
| --- | ---------------- | -------------------------- |
| db  |                  | An operation on a database |
|     | db.connection    |                            |
|     | db.transaction   |                            |
|     | db.sql.query     |                            |

> |     | db.query         |                            |

Reviewer 2: Yes.

|     | db.query.compile |                            |

## Serverless (FAAS)

| Op  | Usage       | Description |
| --- | ----------- | ----------- |
| aws |             |             |
|     | aws.lambda  |             |
|     | aws.request |             |

> | gcp |             |             |

Author: You think they should follow `http.server` possibly? We could also special case `serverless`.

It's important to keep categories consistent between sdks and integrations as they are used by Sentry in the [Operations Breakdown](https://docs.sentry.io/product/sentry-basics/tracing/event-detail/#operations-breakdown) feature. For example, both `db.init` and `db.query` will be categorized as database operations. The default operations breakdown config can be seen [here](https://github.com/getsentry/sentry/blob/809b7fe54c6f06cc1e4c503cf83ded896472a011/src/sentry/projectoptions/defaults.py#L74).

Having both a category and action is recommended. Operations should be lowercased.

# List of Operations

## Browser

| Category    | Usage           | Description                                                                                                         |
| ----------- | --------------- | ------------------------------------------------------------------------------------------------------------------- |

> | pageload    |                 | A full page load of a web application                                                                               |

Reviewer 3: Should we specify what should go into the description of the spans as well? I think it would be awesome to specify them precisely, so these are consistent as well.

Author: Yes I think this is a great idea, but out of the scope of this PR. I want to focus on getting the operations right, we can then take a look at descriptions next.

Reviewer 3: I created an issue for that https://github.com/getsentry/develop/issues/463

|          | db.query         |                            |
|          | db.query.compile |                            |

## Serverless (FAAS)

| Category | Usage            | Description |
| -------- | ---------------- | ----------- |
| faas     |                  |             |
|          | faas.aws         |             |
|          | faas.aws.lambda  |             |
|          | faas.aws.request |             |
|          | faas.gcp         |             |

## Mobile

| Category   | Usage          | Description                         |
| ---------- | -------------- | ----------------------------------- |
| app        |                | Data about the mobile app           |
|            | app.start      |                                     |
|            | app.start.warm |                                     |
|            | app.start.cold |                                     |
| ui         |                | An operation on a mobile/desktop UI |
|            | ui.load        |                                     |
|            | ui.action      |                                     |
| navigation |                | Navigating to another screen        |

> | file       |                | Operations on the file system       |

Reviewer 3: In a call, we discussed what to put in the description of the file span. We thought that filename + bytes count `file.txt (203 KB)` should go in the description and the full file path should go into data of the span as `data["filepath"] = "/absolute/example/file.txt"`. Should the bytes count be KB, MB, GB, ... ?

Can we document this here?

Author: > Can we document this here?

We should - but I'm not going to include it in this PR.

Instead we should be trying to develop our own set of semantic conventions (especially for desktop/mobile/browser - areas where open telemetry doesn't have anything). I plan to take this on after this PR gets merged in.

Reviewer 3: Okay, I created an issue so this doesn't get lost https://github.com/getsentry/develop/issues/463

---

> title: "Span Operations"

Author: Opened https://github.com/getsentry/develop/issues/464 to track that

---

> Span operations are a short code identifying the type of operation the span is measuring. Span operations are low cardinality attributes - they should be as general as possible while still being human readable and useful. They should avoid including high cardinality data like ids and URLs.

Reviewer: ```suggestion
Span operations are a short code identifying the type of operation the span is measuring. Span operations are low cardinality attributes - they should be as general as possible while still being human readable and useful. They should avoid including high cardinality data like IDs and URLs.
```

nit: We use `ID`/`IDs` in prose here and in docs.sentry.io.

Operations are expected to follow [OpenTelemetry's semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/README.md) as much as possible.

Below is a table containing a list of span operations that are recommended to be used.

Operations are expected to generally take the form: `CATEGORY.ACTION`, where the categories and actions can have optional identifiers. For example, a redis database operation takes the form `db.sql.query`, where `db` is the category, `sql` is the category identifier, and `query` is the action. Another example is a mobile app rendering it's screen. It would have the operation `ui.load.hot`, where `ui` is the category, `load` is the action, and `hot` is the action identifier.

> It's important to keep categories consistent between sdks and integrations as they are used by Sentry in the [Operations Breakdown](https://docs.sentry.io/product/sentry-basics/tracing/event-detail/#operations-breakdown) feature. For example, both `db.init` and `db.query` will be categorized as database operations (`db`). The default operations breakdown config can be seen [here](https://github.com/getsentry/sentry/blob/809b7fe54c6f06cc1e4c503cf83ded896472a011/src/sentry/projectoptions/defaults.py#L74).

Reviewer: ```suggestion
It's important to keep categories consistent between SDKs and integrations as they are used by Sentry in the [Operations Breakdown](https://docs.sentry.io/product/sentry-basics/tracing/event-detail/#operations-breakdown) feature. For example, both `db.init` and `db.query` will be categorized as database operations (`db`). The default operations breakdown config can be seen [here](https://github.com/getsentry/sentry/blob/809b7fe54c6f06cc1e4c503cf83ded896472a011/src/sentry/projectoptions/defaults.py#L74).
```

> Due to use of categories in the Sentry product, span categories are expected to be low cardinality data. **Categories should not be auto-generated**. SDK developers should feel free to automatically generate the actions and identifiers for span operations.

Reviewer: Needs verification

- [ ] We start out saying "Span operations are low cardinality attributes", and this sentence here relaxes that expectation, essentially accepting high-cardinality span operations, which sounds contradictory at a minimum, and might have negative performance consequences.

Reviewer 4: The important point here isn't that it's generated, but that the span categories stay low cardinality. I think just recommending that if a span category **is** generated, that it should only use low cardinality inputs (a static array like how we create the ops breakdown measurements) and not something like an `id`, `url`, etc. 

Since `generated` is a fairly conflated term, I'd perhaps rewrite this to state that:
- It should be low cardinality
- If you **are** generating this somehow, take care to not use high cardinality inputs

| serialize |                 | Serialization of data                                              |
| console   |                 | Accessing web servers through the command line (ex. Rails console) |

## Web Frameworks

| Category | Usage             | Description |
| -------- | ----------------- | ----------- |
| django   |                   |             |
|          | django.middleware |             |
|          | django.view       |             |
| express  |                   |             |
| rails    |                   |             |
| rack     |                   |             |

## Database

Databased related spans are expected to follow OpenTelemetry's [Database](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md) semantic conventions when possible.

| Category | Usage            | Description                |
| -------- | ---------------- | -------------------------- |
| db       |                  | An operation on a database |
|          | db.connection    |                            |
|          | db.transaction   |                            |

> |          | db.sql.query     |                            |

Reviewer 5: Still a bit unclear what's the diff between `db.sql.query` and `db.query`? Should we just get rid of `db.sql.query`?

Author: According to the outlined spec `sql` is an identifier for the `db` category. I'm of the opinion that both are fine to exist - `db.query` is just the more general one. 

The key takeaway here is that since both use the same category, `db`, they will be classified the same way in the product. Everything after the `db` prefix just provides additional context, SDK developers should feel free to choose whatever they think fits the best (as long as the category stays consistent).

Open to discussing further if you feel that we should take a more opinionated approach.

Reviewer 5: Ah alright, I thought it will affect the operation breakdown rendering in the UI, but if it's just there for context, I'm good with it, thanks

|            | resource.img    |                                                                                                                     |
| browser    |                 | Usage of browser APIs or functionality                                                                              |
|            | browser.paint   |                                                                                                                     |
| mark       |                 | Usage of [performance.mark() API](https://developer.mozilla.org/en-US/docs/Web/API/Performance/mark)                |

> | measure    |                 | Usage of [performance.measure() API](https://developer.mozilla.org/en-US/docs/Web/API/Performance/measure)          |

Reviewer: Hmm I feel like I could live without this commit; but you adding it to this PR suggests the distinction between `Category` and `Usage` columns is not really necessary.

All we have as actual operations in SDKs are strings, and since we've dropped the intention to document how categories work we could as well have a single column:

| Operation | Description |
| --- | --- |
| ... | ... |

Author: Will address this in the next PR.
